lib/corefoundation.rb

```diff
@@ -1,20 +1,48 @@
-require 'ffi'
-require 'iconv' if RUBY_VERSION < "1.9"
+require "ffi" unless defined?(FFI)
 
-# CF Types
-require 'corefoundation/base'
-require 'corefoundation/null'
-require 'corefoundation/string'
-require 'corefoundation/array'
-require 'corefoundation/boolean'
-require 'corefoundation/data'
-require 'corefoundation/dictionary'
-require 'corefoundation/number'
-require 'corefoundation/date'
-require 'corefoundation/extensions'
+module CF
+  extend FFI::Library
+  ffi_lib '/System/Library/Frameworks/CoreFoundation.framework/CoreFoundation'
 
-# Error wrapper
-require 'corefoundation/error'
+  if FFI::Platform::ARCH == 'x86_64'
+    typedef :long_long, :cfindex
+    typedef :long_long, :cfcomparisonresult
+    typedef :ulong_long, :cfoptionflags
+    typedef :ulong_long, :cftypeid
+    typedef :ulong_long, :cfhashcode
+  else
+    typedef :long, :cfindex
+    typedef :long, :cfcomparisonresult
+    typedef :ulong, :cfoptionflags
+    typedef :ulong, :cftypeid
+    typedef :ulong, :cfhashcode
+  end
 
-# Utilities
-require 'corefoundation/preferences'
+  typedef :pointer, :cftyperef
+end
+
+require_relative 'corefoundation/memory'
+require_relative 'corefoundation/register'
+require_relative 'corefoundation/base'
+require_relative 'corefoundation/null'
+require_relative 'corefoundation/range'
+require_relative 'corefoundation/string'
+require_relative 'corefoundation/array'
+require_relative 'corefoundation/boolean'
+require_relative 'corefoundation/data'
+require_relative 'corefoundation/dictionary'
+require_relative 'corefoundation/number'
+require_relative 'corefoundation/date'
+require_relative 'corefoundation/exceptions'
+require_relative 'corefoundation/preferences'
+require_relative 'corefoundation/extensions'
+
+# REVIEW: Include patches from module
+Integer.include CF::CoreExtensions::Integer
+Float.include CF::CoreExtensions::Float
+Array.include CF::CoreExtensions::Array
+TrueClass.include CF::CoreExtensions::TrueClass
+FalseClass.include CF::CoreExtensions::FalseClass
+String.include CF::CoreExtensions::String
+Time.include CF::CoreExtensions::Time
+Hash.include CF::CoreExtensions::Hash
```

Comment:
You should really consider using refinements here instead of moneypatching in extension methods that are visible to every user of this gem.  The `#to_cf` method might not collide with anything else, but all it would take is something like a CloudFoundry related gem to decide to do that as well.  The the `String#binary?` and `String#binary!` methods could very easily collide with some other gem out there (or with future versions of ruby).  Best would be to remove the need to do this entirely.

Reply:
Yeah  I've had mixed feelings for this code from the start. Didn't get any solid responses on it so just refactored this. Thanks for the review! I'll update this in a separate PR.

Reply:
Update in #23

Reply:
Linking the discussion regarding this for context, the previous repo owner suggested an alternative, i.e. refinements
https://github.com/chef/corefoundation/pull/10#discussion_r679433874

@rishichawda we have to get the `refinements` approach reviewed from Lamont